kad-dht/README.md

```diff
@@ -0,0 +1,288 @@
+
+For recap, `dXOR(sha256(id1), sha256(id2))` is the number of common leftmost
+bits between SHA256 of each peer IDs. The `dXOR` between us and a peer X
+designates the bucket index that peer X will take up in the Kademlia routing
+table.
+
+## Kademlia routing table
+
+The data structure backing this system is a k-bucket routing table, closely
+following the design outlined in the Kademlia paper [0]. The default value for
+`k` is 20, and the maximum bucket count matches the size of the SHA256 function,
+i.e. 256 buckets.
+
+The routing table is unfolded lazily, starting with a single bucket a position 0
+(representing the most distant peers), and splitting it subsequently as closer
+peers are found, and the capacity of the nearmost bucket is exceeded.
+
+## Alpha concurrency factor (α)
+
+The concurrency of node and value lookups are limited by parameter `α`, with a
+default value of 3. This implies that each lookup process can perform no more
+than 3 inflight requests, at any given time.
+
+## Record keys
+
+Records in the DHT are keyed by CID [4]. There are intentions to move to
+multihash [5] keys in the near future, as certain CID components like the
+multicodec are redundant.
+
+## Interfaces
+
+The libp2p Kad DHT implementation satisfies the routing interfaces:
+
+```go
+type Routing interface {
+	ContentRouting
+	PeerRouting
+	ValueStore
+
+	// Kicks off the bootstrap process.
+	Bootstrap(context.Context) error
+}
+
+// ContentRouting is used to find information about who has what content.
+type ContentRouting interface {
+	// Provide adds the given CID to the content routing system. If 'true' is
+	// passed, it also announces it, otherwise it is just kept in the local
+	// accounting of which objects are being provided.
+	Provide(context.Context, cid.Cid, bool) error
+
+	// Search for peers who are able to provide a given key.
+	FindProvidersAsync(context.Context, cid.Cid, int) <-chan pstore.PeerInfo
+}
+
+// PeerRouting is a way to find information about certain peers.
+//
+// This can be implemented by a simple lookup table, a tracking server,
+// or even a DHT (like herein).
+type PeerRouting interface {
+	// FindPeer searches for a peer with given ID, returns a pstore.PeerInfo
+	// with relevant addresses.
+	FindPeer(context.Context, peer.ID) (pstore.PeerInfo, error)
+}
+
+// ValueStore is a basic Put/Get interface.
+type ValueStore interface {
+	// PutValue adds value corresponding to given Key.
+	PutValue(context.Context, string, []byte, ...ropts.Option) error
+
+	// GetValue searches for the value corresponding to given Key.
+	GetValue(context.Context, string, ...ropts.Option) ([]byte, error)
+}
+```
+
+## Value lookups
+
+When looking up an entry in the DHT, the implementor should collect at least `Q`
+(quorum) responses from distinct nodes to check for consistency before returning
+an answer.
+
+Should the responses be different, the `Validator.Select()` function is used to
+resolve the conflict and select the _best_ result.
```

Comment:
We should notice here that the `Validator.Select()` is associated with a specific namespace [libp2p/go-libp2p-record/validator.go#L56](https://github.com/libp2p/go-libp2p-record/blob/master/validator.go#L56). 

Moreover, I have been thinking about this for a while in JS land. In case we do not have a `Validator.Select()` for the namespace of a key being used, shouldn't we fallback to a default `Select` function? In JS land, when using arbitrary keys with unknown namespaces it failed to get the record. We ended up changing to selecting the first record, but I don't know if that is the best approach.

I believe the same happens with `Validate()`.

Reply:
Can validation be deferred to consumers of the DHT? It's not really a requirement to participate in it?

Reply:
It's not strictly necessary, no. But it would be nice if nodes can throw out clearly bogus records instead of storing them. So this should be suggested but optional for now.

Reply:
That is, it would be nice if the node that receives a `PUT_VALUE` can do some sanity checking.

Reply:
Hm, per the [comment on the refactor proposal](https://github.com/libp2p/research-dht/pull/8#issuecomment-453289582), the DHT node implementation could call a registered handler on receiving a `PUT_VALUE`, which does whatever it wishes with the data.

Reply:
Yes, that's the idea. We could move this validation outside the DHT itself.

Reply:
https://github.com/libp2p/specs/pull/108/commits/6909f8a97685c3f60679dbe7c9e81a5ab411bcf7 rewords the validation description, allowing implementations to use any mechanism for validation, including validation through consumers of the DHT. I am resolving this conversation for now. Let me know if you would prefer further adjustments of the section.